streaming/src/main/scala/org/apache/spark/streaming/scheduler/StreamingListenerBus.scala

```diff
@@ -64,36 +71,40 @@ private[spark] class StreamingListenerBus() extends Logging {
   }
 
   def addListener(listener: StreamingListener) {
-    listeners += listener
+    listeners.add(listener)
   }
 
   def post(event: StreamingListenerEvent) {
+    if (stopped) {
+      // Drop further events to make `StreamingListenerShutdown` be delivered ASAP
+      logError("StreamingListenerBus has been stopped! Drop " + event)
+      return
+    }
     val eventAdded = eventQueue.offer(event)
-    if (!eventAdded && !queueFullErrorMessageLogged) {
+    if (!eventAdded && queueFullErrorMessageLogged.compareAndSet(false, true)) {
       logError("Dropping StreamingListenerEvent because no remaining room in event queue. " +
         "This likely means one of the StreamingListeners is too slow and cannot keep up with the " +
         "rate at which events are being started by the scheduler.")
-      queueFullErrorMessageLogged = true
     }
   }
 
-  /**
-   * Waits until there are no more events in the queue, or until the specified time has elapsed.
-   * Used for testing only. Returns true if the queue has emptied and false is the specified time
-   * elapsed before the queue emptied.
-   */
-  def waitUntilEmpty(timeoutMillis: Int): Boolean = {
-    val finishTime = System.currentTimeMillis + timeoutMillis
-    while (!eventQueue.isEmpty) {
-      if (System.currentTimeMillis > finishTime) {
-        return false
+  def stop(): Unit = {
+    stopped = true
+    // Should not call `post`, or `StreamingListenerShutdown` may be dropped.
+    eventQueue.put(StreamingListenerShutdown)
```

Comment:
I would prefer this to make this bus use the new behavior in `LiveListenerBus` as well. It's a little confusing that this one keeps using a shutdown event and joins on the thread instead. I don't see a particular requirement in the `StreamingListenerBus` that prevents us from doing this.